llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp

```diff
@@ -56,6 +57,10 @@ class RISCVVLOptimizer : public MachineFunctionPass {
   std::optional<MachineOperand> checkUsers(MachineInstr &MI);
   bool tryReduceVL(MachineInstr &MI);
   bool isCandidate(const MachineInstr &MI) const;
+
+  /// For a given instruction, records what elements of it are demanded by
+  /// downstream users.
+  DenseMap<const MachineInstr *, MachineOperand> DemandedVLs;
```

Comment:
Please just have the value be an optional<MachineOperand>.  This simplifies the code, and has no real additional cost.  Please stop trying to micro-optimize this.  If you want to come back to it, please feel free, but let's do the simple version first.  

To be clear, this is *not* just a stylistic point.  I am suspecting from your code structure below that you have a bug in your handling of the std::nullopt case, and want as much of that logic to disappear from this review as possible.

Reply:
More than happy to do so! Hope this didn't come across as an optimisation/stylistic point, that wasn't my intention.